### PR TITLE
fix: send aggregated response bump

### PR DESCRIPTION
**Pull request by MarcosNicolau**
## Description

Fixes aggregator bump fee by adding the following checks:
- The newly bumped gas price is at least `10%` higher than the previously sent transaction.
- If the receipt wait timeout for a tx has passed, save the tx.
- On the next iteration, before sending the transaction, loop through the vector and try to get the receipt again.
- If the receipt is not found, check if the batch is marked as responded in the contract
- If all that fails, send the transaction again with the newly bumped gasPrice.

Finally, tracing has been improved by adding step-by-step logs with the respective merkle root hash.

## Type of change

- [x] Bug fix

## Checklist

- [ ] “Hotfix” to `testnet`, everything else to `staging`
- [ ] Linked to Github Issue
- [ ] This change depends on code or research by an external entity
  - [ ] Acknowledgements were updated to give credit
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run
- [ ] Has a known issue
  - [Link to the open issue addressing it]() 
- [ ] If your PR changes the Operator compatibility (Ex: Upgrade prover versions)
  - [ ] This PR adds compatibility for operator for both versions and do not change batcher/docs/examples
  - [ ] This PR updates batcher and docs/examples to the newer version. This requires the operator are already updated to be compatible

### CODE REVIEW

File: core/chainio/avs_writer.go [hunk 1]
```diff
@@ -81,69 +82,112 @@ func NewAvsWriterFromConfig(baseConfig *config.BaseConfig, ecdsaConfig *config.E
...
+		// if the suggested bump does not meet the minimum threshold, use a fallback calculation to slightly increment the previous gas price.
+		if suggestedBumpedGasPrice.Cmp(minimumGasPriceBump) > 0 {
+			txOpts.GasPrice = suggestedBumpedGasPrice
 		} else {
-			// bump the last tx gas price a little by `gasBumpIncrementalPercentage` to replace it.
-			txOpts.GasPrice = utils.CalculateGasPriceBumpBasedOnRetry(txOpts.GasPrice, gasBumpIncrementalPercentage, 0, 0)
+			txOpts.GasPrice = utils.CalculateGasPriceBumpBasedOnRetry(
+				previousTxGasPrice,
+				gasBumpPercentage,
+				gasBumpIncrementalPercentage,
```
Comment (uri-99) [inline core/chainio/avs_writer.go]: why not simply set the already calculated `minimumGasPriceBump` ?

File: core/chainio/avs_writer.go [hunk 2]
```diff
@@ -81,69 +82,112 @@ func NewAvsWriterFromConfig(baseConfig *config.BaseConfig, ecdsaConfig *config.E
...
+				i,
+			)
 		}
 
 		if i > 0 {
+			w.logger.Infof("Trying to get old sent transaction receipt before sending a new transaction", "merkle root", batchMerkleRootHashString)
+			for _, tx := range sentTxs {
+				receipt, _ := w.Client.TransactionReceipt(context.Background(), tx.Hash())
+				if receipt != nil {
```
Comment (uri-99) [inline core/chainio/avs_writer.go]: if err != nil ?
or receipt == nil ?

File: core/chainio/avs_writer.go [hunk 3]
```diff
@@ -81,69 +82,112 @@ func NewAvsWriterFromConfig(baseConfig *config.BaseConfig, ecdsaConfig *config.E
...
+						w.checkIfAggregatorHadToPaidForBatcher(tx, batchIdentifierHash)
+						return receipt, nil
+					}
+				}
+			}
+			w.logger.Infof("Receipts for old transactions not found, will check if the batch state has been responded", "merkle root", batchMerkleRootHashString)
+			batchState, _ := w.BatchesStateRetryable(&bind.CallOpts{}, batchIdentifierHash)
+			if batchState.Responded {
+				w.logger.Infof("Batch state has been already responded", "merkle root", batchMerkleRootHashString)
+				return nil, nil
```
Comment (uri-99) [inline core/chainio/avs_writer.go:150]: a comment that nil,nil means OK would be good

File: core/chainio/avs_writer.go [hunk 4]
```diff
@@ -81,69 +82,112 @@ func NewAvsWriterFromConfig(baseConfig *config.BaseConfig, ecdsaConfig *config.E
...
 			return nil, err
 		}
 		return nil, fmt.Errorf("transaction failed")
 	}
 
-	return retry.RetryWithData(respondToTaskV2Func, retry.MinDelay, retry.RetryFactor, 0, retry.MaxInterval, 0)
+	// This just retries the bump of a fee in case of a timeout
+	// The wait is done before on WaitForTransactionReceiptRetryable, and all the functions are retriable,
+	// so this retry doesn't need to wait more time
+	maxInterval := time.Millisecond * 500
```
Comment (uri-99) [inline core/chainio/avs_writer.go:194]: magic number

File: core/chainio/avs_writer.go [hunk 5]
```diff
@@ -169,9 +213,23 @@ func (w *AvsWriter) checkIfAggregatorHadToPaidForBatcher(tx *types.Transaction,
 }
 
 func (w *AvsWriter) checkAggAndBatcherHaveEnoughBalance(tx *types.Transaction, txOpts bind.TransactOpts, batchIdentifierHash [32]byte, senderAddress [20]byte) error {
+	defer func() {
+		if r := recover(); r != nil {
+			w.logger.Error("Recovered from panic", "error", r)
+		}
+		// return fmt.Errorf("Recovered from panic") // TODO can't do this
```
Comment (uri-99) [inline core/chainio/avs_writer.go]: remove comment

File: core/chainio/avs_writer.go [hunk 6]
```diff
@@ -169,9 +213,23 @@ func (w *AvsWriter) checkIfAggregatorHadToPaidForBatcher(tx *types.Transaction,
...
 	aggregatorAddress := txOpts.From
-	txCost := new(big.Int).Mul(new(big.Int).SetUint64(tx.Gas()), txOpts.GasPrice)
+	txGasAsBigInt := new(big.Int).SetUint64(tx.Gas())
+	txGasPrice := txOpts.GasPrice
+	w.logger.Info("Transaction Gas Cost", "cost", txGasAsBigInt)
+	w.logger.Info("Transaction Gas Price", "cost", txGasPrice)
+
+	txCost := new(big.Int).Mul(txGasAsBigInt, txGasPrice)
+
+	// txCost := new(big.Int).Mul(new(big.Int)tx.Gas()), txOpts.GasPrice)
```
Comment (uri-99) [inline core/chainio/avs_writer.go]: remove comment

**Review by uri-99 — CHANGES_REQUESTED** [review]

File: config-files/config-aggregator.yaml [hunk 7]
```diff
@@ -38,9 +38,10 @@ aggregator:
   garbage_collector_tasks_age: 20 #The age of tasks that will be removed by the GC, in blocks. Suggested value for prod: '216000' (30 days)
   garbage_collector_tasks_interval: 10 #The interval of queried blocks to get an old batch. Suggested value for prod: '900' (3 hours)
   bls_service_task_timeout: 168h # The timeout of bls aggregation service tasks. Suggested value for prod '168h' (7 days)
-  gas_base_bump_percentage: 10 # How much to bump gas price when responding to task. Suggested value 10%
-  gas_bump_incremental_percentage: 2 # An extra percentage to bump every retry i*2 when responding to task. Suggested value 2%
-  time_to_wait_before_bump: 36s # The time to wait for the receipt when responding to task. Suggested value 36 seconds (3 blocks)
+  gas_base_bump_percentage: 25 # Percentage to overestimat gas price when sending a task
```
Comment (JuArce) [inline config-files/config-aggregator.yaml]:
```suggestion
  gas_base_bump_percentage: 25 # Percentage to overestimate gas price when sending a task
```

File: config-files/config-aggregator.yaml [hunk 8]
```diff
@@ -38,9 +38,10 @@ aggregator:
...
+  gas_bump_incremental_percentage: 20 # An extra percentage to overestimate in each bump of respond to task. This is additive between tries
+  #Gas used formula = est_gas_by_node * (gas_base_bump_percentage + gas_bum_incremental_percentage * i) / 100, where i is the iteration number.
```
Comment (JuArce) [inline config-files/config-aggregator.yaml]:
```suggestion
  # Gas used formula = est_gas_by_node * (gas_base_bump_percentage + gas_bum_incremental_percentage * i) / 100, where i is the iteration number.
```

File: config-files/config-aggregator.yaml [hunk 9]
```diff
@@ -38,9 +38,10 @@ aggregator:
...
+  time_to_wait_before_bump: 72s # The time to wait for the receipt when responding to task. Suggested value 36 seconds (3 blocks)
```
Comment (JuArce) [inline config-files/config-aggregator.yaml]:
```suggestion
  time_to_wait_before_bump: 72s # The time to wait for the receipt when responding to task. Suggested value 72 seconds (6 blocks)
```

File: core/chainio/avs_writer.go [hunk 10]
```diff
@@ -93,39 +94,78 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
...
 		if err != nil {
 			return nil, err
 		}
-
-		bumpedGasPrice := utils.CalculateGasPriceBumpBasedOnRetry(gasPrice, gasBumpPercentage, gasBumpIncrementalPercentage, i)
-		// new bumped gas price must be higher than the last one (this should hardly ever happen though)
-		if bumpedGasPrice.Cmp(txOpts.GasPrice) > 0 {
-			txOpts.GasPrice = bumpedGasPrice
+		previousTxGasPrice := txOpts.GasPrice
+		// minimum gas price required for the new transaction to ensure it's 10% higher than the previous one.
```
Comment (JuArce) [inline core/chainio/avs_writer.go]: Is this outdated?

**Review by JuArce — CHANGES_REQUESTED** [review]

File: core/chainio/avs_writer.go (hunk 3 above)
Comment (MarcosNicolau) [reply core/chainio/avs_writer.go:150]: I've added a description in the function at the top that explains this behavior.

File: core/chainio/avs_writer.go (hunk 10 above)
Comment (MarcosNicolau) [reply core/chainio/avs_writer.go]: No, it was not, but the comment was a little confusing. I've updated it, it should be clearer now.

**Review by JuArce — APPROVED** [review]

**Review by uri-99 — APPROVED** [review]

**Review by MauroToscano — CHANGES_REQUESTED** [review]
In review by Pelito

**Review by Oppen — APPROVED** [review]
LGTM.
Two issues to file:
1. Sequential requests quadratic to the number of attempts. This can be postponed because it requires several attempts to be an issue, potentially many, so we're likely under a more problematic scenario than some slowness already.
2. Metrics for the number of attempts and tracing would be more desirable than just logging (in fact, if we need to choose only one, tracing is better). This include events and spans, but can be a later refinement as well. That will help with diagnosis if this becomes a bottleneck.

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**